java/source/data.rst

```diff
@@ -0,0 +1,360 @@
+    TypeEqualsVisitor visitor = new TypeEqualsVisitor(right);
+
+    System.out.println(visitor.equals(left1));
+    System.out.println(visitor.equals(left2));
+
+.. testoutput::
+
+    true
+    false
+
+Compare Values on the Array
+===========================
+
+Comparing two values at the given indices in the vectors:
+
+.. testcode::
+
+    import org.apache.arrow.algorithm.sort.StableVectorComparator;
+    import org.apache.arrow.algorithm.sort.VectorValueComparator;
+    import org.apache.arrow.vector.VarCharVector;
+    import org.apache.arrow.memory.RootAllocator;
+
+    class TestVectorValueComparator extends VectorValueComparator<VarCharVector> {
+        @Override
+        public int compareNotNull(int index1, int index2) {
+            byte b1 = vector1.get(index1)[0];
+            byte b2 = vector2.get(index2)[0];
+            return b1 - b2;
+        }
+
+        @Override
+        public VectorValueComparator<VarCharVector> createNew() {
+            return new TestVectorValueComparator();
+        }
+    }
+
+    RootAllocator rootAllocator = new RootAllocator(Long.MAX_VALUE);
+
+    VarCharVector vec = new VarCharVector("valueindexcomparator", rootAllocator);
+    vec.allocateNew(3);
+    vec.setValueCount(3);
+    vec.set(0, "ba".getBytes());
+    vec.set(1, "abc".getBytes());
+    vec.set(2, "aa".getBytes());
+
+    VectorValueComparator<VarCharVector> comparatorValues = new TestVectorValueComparator();
+    VectorValueComparator<VarCharVector> stableComparator = new StableVectorComparator<>(comparatorValues);
+    stableComparator.attachVector(vec);
+
+    System.out.println(stableComparator.compare(0, 1) > 0);
+    System.out.println(stableComparator.compare(1, 2) < 0);
+
+.. testoutput::
+
+    true
+    true
+
+Search Values on the Array
+==========================
+
+Linear Search - O(n)
+********************
+
+Algorithm: org.apache.arrow.algorithm.search.VectorSearcher#linearSearch - O(n)
+
+.. testcode::
+
+    import org.apache.arrow.algorithm.search.VectorSearcher;
+    import org.apache.arrow.algorithm.sort.DefaultVectorComparators;
+    import org.apache.arrow.algorithm.sort.VectorValueComparator;
+    import org.apache.arrow.vector.IntVector;
+    import org.apache.arrow.memory.RootAllocator;
+
+    RootAllocator rootAllocator = new RootAllocator(Long.MAX_VALUE);
+    IntVector linearSearchVector = new IntVector("linearSearchVector", rootAllocator);
+    linearSearchVector.allocateNew(10);
+    linearSearchVector.setValueCount(10);
+    for (int i = 0; i < 10; i++) {
+        linearSearchVector.set(i, i);
+    }
+    VectorValueComparator<IntVector> comparatorInt = DefaultVectorComparators.createDefaultComparator(linearSearchVector);
+    List<Integer> listResultLinearSearch = new ArrayList<Integer>();
+    for (int i = 0; i < 10; i++) {
+       int result = VectorSearcher.linearSearch(linearSearchVector, comparatorInt, linearSearchVector, i);
+       listResultLinearSearch.add(result);
+    }
```

Comment:
I think that the for loop to search for all values makes things more complicate to understand. Let's stick to how you search 1 value, it will make the recipe easier to understand.

Reply:
Changed